### PR TITLE
Hash interface added, support for multiple hash function to follow in few weeks.

### PR DESCRIPTION
**Pull request by hisenberg08**
*Please fill in the fields below to submit a pull request.  The more
information that is provided, the better.*

**Fixes issue #**:
A generic hash interface add, which could be extended for multiple hash functions. Right now it is only working for sha256 but support has been added for sha512 and md5. To completely make them work test cases have to be modified. I would be working on adding test cases for multiple hash functions in coming weeks.

**Description of pull request**:

**Please verify and check that the pull request fulfills the following
requirements**:

- [] Tests have been added for the bug fix or new feature
- [] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by SantiagoTorres** [comment]
Thanks @hisenberg08 !

> /home/travis/.travis/functions: line 104: test: in_toto/hash.go: binary operator expected

Hmm, it appears the tests aren't working properly due to some formatting issue:

```bash
santiago at ~/.../in-toto/in-toto-golang ✔ gofmt -d in_toto
```

```diff
diff -u in_toto/hash.go.orig in_toto/hash.go
--- in_toto/hash.go.orig	2019-04-08 16:14:22.181004158 -0400
+++ in_toto/hash.go	2019-04-08 16:14:22.181004158 -0400
@@ -1,62 +1,54 @@
 package in_toto
 
 import (
-
-	"fmt"
+	"crypto/md5"
 	"crypto/sha256"
 	"crypto/sha512"
-	"crypto/md5"
-
+	"fmt"
 )
 
-
 /*
 Hash is an interface which contains a generic compute method.
 This method is implemented by several functions to return the
 hash for a given data.
 
 To add hash functions which are currently not supported one
-has to create a struct, write it's defination, add it's 
-value to the map in createMap function and add the hash 
+has to create a struct, write it's defination, add it's
+value to the map in createMap function and add the hash
 function to the list of hash functions(if required)
 */
 
 type Hash interface {
-
-	Compute(content []uint8) (string)
+	Compute(content []uint8) string
 }
 
-
-
 /*
 Declaration of struct, one for each of the hash function.
 */
-type sha_256_Hash struct {}
-type sha_512_Hash struct {}
-type md5_Hash struct {}
-
+type sha_256_Hash struct{}
+type sha_512_Hash struct{}
+type md5_Hash struct{}
 
 /*
 Defination of compute function for each of the hash struct
 declared above.
 */
 
-func (hash *sha_256_Hash) Compute(content []uint8) (string){
-	
-	hashed := sha256.Sum256(content);
+func (hash *sha_256_Hash) Compute(content []uint8) string {
+
+	hashed := sha256.Sum256(content)
 	n := fmt.Sprintf("%x", hashed)
 	return n
 }
 
-
-func (hash *sha_512_Hash) Compute(content []uint8) (string){
-	hashed := sha512.Sum512(content);
+func (hash *sha_512_Hash) Compute(content []uint8) string {
+	hashed := sha512.Sum512(content)
 	n := fmt.Sprintf("%x", hashed)
 	return n
 }
 
-func (hash *md5_Hash) Compute(content []uint8) (string){
-	hashed := md5.Sum(content);
+func (hash *md5_Hash) Compute(content []uint8) string {
+	hashed := md5.Sum(content)
 	n := fmt.Sprintf("%x", hashed)
 	return n
 }
@@ -66,13 +58,13 @@
 their respective reference object as value.
 */
 
-func createMap() map[string]interface{Compute(content []uint8) (string)}{
-	mapper := map[string]interface{Compute(content []uint8) (string)}{
-	        "sha256"  : &sha_256_Hash{},
-	        "sha512"  : &sha_512_Hash{},
-	        "md5"  	  : &md5_Hash{},
-		}
-		return mapper
+func createMap() map[string]interface{ Compute(content []uint8) string } {
+	mapper := map[string]interface{ Compute(content []uint8) string }{
+		"sha256": &sha_256_Hash{},
+		"sha512": &sha_512_Hash{},
+		"md5":    &md5_Hash{},
+	}
+	return mapper
 }
 
 /*
@@ -80,7 +72,7 @@
 for each of the files.
 */
 
-func createList() []string{
+func createList() []string {
 
 	hash_func := []string{"sha256"}
 	return hash_func
diff -u in_toto/runlib.go.orig in_toto/runlib.go
--- in_toto/runlib.go.orig	2019-04-08 16:14:22.191004192 -0400
+++ in_toto/runlib.go	2019-04-08 16:14:22.191004192 -0400
@@ -24,12 +24,12 @@
 //is the reason for code failure. It will be removed once the code is completed.
 
 func RecordArtifact(path string) (map[string]interface{}, error) {
-	
+
 	mapper := createMap()
 
 	// Read file from passed path
 	content, err := ioutil.ReadFile(path)
-	
+
 	retMap := make(map[string]interface{})
 
 	if err != nil {
@@ -39,7 +39,7 @@
 	// Create a map of all the hashes present in the hash_func list
 	hash_func := createList()
 	for _, element := range hash_func {
-		retMap[element] = mapper[element].Compute([] uint8(content))
+		retMap[element] = mapper[element].Compute([]uint8(content))
 	}
 
 	// Return it in a format that is conformant with link metadata artifacts
```

Could you please fix these? Thanks!

File: in_toto/hash.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	"fmt"
+)
+
+/*
+Hash is an interface which contains a generic compute method.
+This method is implemented by several functions to return the
+hash for a given data.
+
+To add hash functions which are currently not supported one
+has to create a struct, write it's defination, add it's
```
Comment (SantiagoTorres) [inline in_toto/hash.go]: I notice that the typo "defination" in the codebase in a couple of places. Could you correct it please?

File: in_toto/hash.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+type Hash interface {
+	Compute(content []uint8) string
+}
+
+/*
+Declaration of struct, one for each of the hash function.
+*/
+type sha_256_Hash struct{}
+type sha_512_Hash struct{}
```
Comment (SantiagoTorres) [inline in_toto/hash.go]: I think it'd be better to name these `sha256Hash` or so.

File: in_toto/hash.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+type md5_Hash struct{}
+
+/*
+Defination of compute function for each of the hash struct
+declared above.
+*/
+
+func (hash *sha_256_Hash) Compute(content []uint8) string {
```
Comment (SantiagoTorres) [inline in_toto/hash.go]: Could you rename the argument `content` to either `data` or `contents`?

File: in_toto/hash_test.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	}
+
+	testMap := make(map[string]interface{})
+	for _, element := range hash_func {
+
+		result := mapper[element].Compute([]uint8("Hashing this string to test"))
+		testMap[element] = result
+	}
+
+	if !reflect.DeepEqual(testMap, expected) {
```
Comment (SantiagoTorres) [inline in_toto/hash_test.go]: Do you need deep equal here?

File: in_toto/runlib.go [hunk 5]
```diff
@@ -21,20 +19,31 @@ using sha256 and returns a map in the following format:
 If reading the file fails, the first return value is nil and the second return
 value is the error.
 */
+
 func RecordArtifact(path string) (map[string]interface{}, error) {
+
+	mapper := createMap()
```
Comment (SantiagoTorres) [inline in_toto/runlib.go]: I believe that both `mapper` and `retMap` are not descriptive enough variable names in the context of  `recordArtifact`

File: in_toto/hash_test.go (hunk 4 above)
Comment (hisenberg08) [reply in_toto/hash_test.go]: Made all the requested changed.

Regarding deep equal, I think we will need it.

According to the description of on stack overflow

"it first checks if both maps are nil, then it checks if they have the same length before finally checking to see if they have the same set of (key, value) pairs."

I found similar description on the go source code 
https://golang.org/src/reflect/deepequal.go?s=2432:2441#L98

Please let me know if you still think deep equal is not apt. I can then write a function to do the check.

**Comment by SantiagoTorres** [comment]
Thanks, @hisenberg08 !

**Pull request merged**